### PR TITLE
CMake 1.1 PG: usability and compatibility improvements

### PR DESCRIPTION
**Pull request by RJVB**
### Description

Triggered by trac #62642 this PR introduces a compatibility improvement for debug builds, delegating that support to pg `debug`.

It also includes a few minor usability improvements for building with `ninja` (and its "lite" alternative, `samu`).

 * use a path-style depspec for `ninja` to accept `port:samurai` too
 * use `-l N` instead of `-j N` when building with `ninja` (constrains the CPU
   load instead of the number of build processes, useful to keep the host
   responsive esp. when a build calls `ninja` recursively).

#### Type(s)

- [X] enhancement

### CODE REVIEW

**Comment by RJVB** [comment]
Technically it appears possible to call the option var `cmake.+debugopts` but that's probably not a good idea? :)

No idea why the checks failed, looks like an issue in the testing implementation?!

File: _resources/port1.0/group/cmake-1.1.tcl [hunk 1]
```diff
@@ -85,6 +92,7 @@ default destroot.target             install/fast
 # make sure cmake is available:
 # can use cmake or cmake-devel; default to cmake if not installed
 depends_build-append                path:bin/cmake:cmake
+depends_skip_archcheck-append       cmake
```
Comment (ryandesign) [inline _resources/port1.0/group/cmake-1.1.tcl:95]: The cmake Portfile already specifies `installs_libs no` so using `depends_skip_archcheck` for cmake here should not be necessary.

File: _resources/port1.0/group/cmake-1.1.tcl [hunk 2]
```diff
@@ -146,7 +154,9 @@ proc cmake::handle_generator {option action args} {
             "*Unix Makefiles*" {
                 ui_debug "Selecting the 'Unix Makefiles' generator ($args)"
                 depends_build-delete \
-                                port:ninja
+                                path:bin/ninja:ninja
+                depends_skip_archcheck-delete \
+                                ninja
```
Comment (ryandesign) [inline _resources/port1.0/group/cmake-1.1.tcl:159]: The ninja Portfile already specifies `installs_libs no` so using `depends_skip_archcheck` for ninja here should not be necessary.

File: _resources/port1.0/group/cmake-1.1.tcl [hunk 3]
```diff
@@ -159,10 +169,21 @@ proc cmake::handle_generator {option action args} {
             "*Ninja" {
                 ui_debug "Selecting the Ninja generator ($args)"
                 depends_build-append \
-                                port:ninja
+                                path:bin/ninja:ninja
+                depends_skip_archcheck-append \
+                                ninja
```
Comment (ryandesign) [inline _resources/port1.0/group/cmake-1.1.tcl:174]: The ninja Portfile already specifies `installs_libs no` so using `depends_skip_archcheck` for ninja here should not be necessary.

**Review by ryandesign — COMMENTED** [review]
> Technically it appears possible to call the option var `cmake.+debugopts` but that's probably not a good idea? :)

I agree. I don't think we've used a plus sign in a variable name before and I would not recommend starting to do so now.

> No idea why the checks failed, looks like an issue in the testing implementation?!

https://trac.macports.org/ticket/62071

**Comment by RJVB** [comment]
On Friday April 16 2021 08:47:24 Ryan Schmidt wrote:

>The cmake Portfile already specifies `installs_libs no` so using `depends_skip_archcheck` for cmake here should not be necessary.
>The ninja Portfile already specifies `installs_libs no` so using `depends_skip_archcheck` for ninja here should not be necessary.

Maybe that's the case for `depends_lib` (and even that I'm not so certain) but it's not for `depends_build` (and `depends_run`) in my experience, sadly. I tend to remember when I had to use the skip archcheck trick because I've always forgotten the syntax by the time I need it again ...

**Comment by szhorvat** [comment]
@RJVB A bit off topic, I'm just curious: I see several references to Ninja. Is it possible to set `ninja` as the default build tool for all CMake PG ports?

**Comment by RJVB** [comment]
Not currently, unless you're willing to change the default in the PG itself.

There isn't much point in doing that though; the advantage of building through ninja becomes visible only when you do lots of incremental (re)builds, and (much less so) for very large projects. The latter can select ninja in their Portfile, and if you're doing the former you're probably working *on* a port and can thus do the same selection in its Portfile.

FWIW, in my experience the overhead (memory, page faults, whatever) of "official ninja" is on par with that of make. Roughly put, ninja is faster in rebuilds mostly because it spends less time figuring out which targets to build. `samu` (port:samurai) is a bit better in that aspect but even with that the bottleneck in an initial (or clean) build is simply the compilation itself.

And (since we're off-topic), again in my experience, I find that even for projects I do active development on I prefer make despite the small gain ninja would give me. Because of that bottleneck, but also because it's easier to build just parts of a project when you have Makefiles in the project (build) subdirectories. Saves me from having to figure out and remember how targets are called; just chdir into the build subdir and call `make`...

**Comment by mascguy** [comment]
I'd love to see support added for additional CMake install paths.

For ports `opencv3` and `opencv4`, we're only able to set three via the portgroup. Otherwise, they're configure arguments:

```
# Populate the CMake options currently available
# TODO: Extend CMake portgroup to support more paths
cmake.install_prefix \
                    ${opencv_install_prefix}
cmake_share_module_dir \
                    ${opencv_install_cmake}
cmake.install_rpath \
                    ${opencv_install_lib}

configure.args-append \
[...non-cmake flags removed for readability...]
                    -DCMAKE_INSTALL_LIBEXECDIR=${opencv_install_prefix} \
                    -DCMAKE_INSTALL_BINDIR=${opencv_install_prefix}/bin \
                    -DCMAKE_INSTALL_SBINDIR=${opencv_install_prefix}/sbin \
                    -DCMAKE_INSTALL_SHAREDIR=${opencv_install_prefix}/share \
                    -DCMAKE_INSTALL_INCLUDEDIR=${opencv_install_include} \
                    -DCMAKE_INSTALL_LIBDIR=${opencv_install_lib} \
                    -DCMAKE_INSTALL_NAME_DIR=${opencv_install_lib} \
```

**Comment by RJVB** [comment]
I don't know what others think, but personally I'd say that the goal of this portgroup is not to provide a bit easier syntax for every CMake option that is used in at least one or two Portfiles... 

Each additional option we add requires an option var and a conditional expression (probably in a procedure) to be evaluated at some point because you only want to add the `-DCMAKE_FOO=value` pair when `value` is defined.

The options that exist currently are justified because they are used commonly (often together, for the 2 install path ones) and/or to hide complex syntax that's just a bit too easy to forget.

**Comment by mascguy** [comment]
> I don't know what others think, but personally I'd say that the goal of this portgroup is not to provide a bit easier syntax for every CMake option that is used in at least one or two Portfiles... Each additional option we add requires an option var and a conditional expression (probably in a procedure) to be evaluated at some point because you only want to add the `-DCMAKE_FOO=value` pair when `value` is defined. The options that exist currently are justified because they are used commonly (often together, for the 2 install path ones) and/or to hide complex syntax that's just a bit too easy to forget.

The goal certainly makes sense, and you're right: Adding a bunch of additional options for things without much demand, doesn't make sense.

**Comment by mascguy** [comment]
Rene, would it be possible to add an option to customize the build command? I ask, as some ports have a need to use `gmake`, rather than `make`, and it appears that the 1.1 portgroup presently hard-codes that.

**Comment by RJVB** [comment]
Sure, why not? Not to seem lazy, but why don't you create a PR with what you have in mind? This one seems to remain up only as a sort of repository of ideas...

**Comment by mascguy** [comment]
> Sure, why not? Not to seem lazy, but why don't you create a PR with what you have in mind? This one seems to remain up only as a sort of repository of ideas...

Sounds good, I'll add it to my list.

**Comment by cooljeanius** [comment]
I remember @kurthindenburg wondering about the debug options recently...

**Comment by szhorvat** [comment]
I still think that adding `-DDEBUG` _by default_ is silly, unnecessary and even dangerous. Unlike `NDEBUG`, this is not a standard macro. Project should not be prepared to be able to handle arbitrary defines injected into their compilation process.

The issue that triggered this PR is https://trac.macports.org/ticket/62642

**Comment by reneeotten** [comment]
this is clearly not going anywhere - if there is no actionable change by the end of the week I am going to close this PR

**Comment by mascguy** [comment]
I just rebased against master, and plan to prioritize all of this over the next week or two. So please keep this open.

**Comment by RJVB** [comment]
>As a reminder, what started this was a request to NOT arbitrarily add `-DDEBUG` whenever using the `+debug` variant. Unlike `NDEBUG`, this is not a standard macro and it is not predictable what its effects will be on various projects. 

FWIW, defining NDEBUG or undefining it is just as unpredictable in terms of how various projects will react to it.

So if you really want to do this properly there would have to be an option to define either, or maybe even better, a variable that defines the token to be defined for non-debug builds and the one for debug builds.

IMHO a +debug variant only makes sense if it defines the full monty to get the utmost debuggability - otherwise the user can just use `configure.optflags=-g` on the commandline.

**Comment by szhorvat** [comment]
> FWIW, defining NDEBUG or undefining it is just as unpredictable in terms of how various projects will react to it.

This is not correct. `NDEBUG` is part of the C standard with a very specific meaning. `DEBUG` is not. https://en.cppreference.com/w/c/error/assert

File: _resources/port1.0/group/cmake-1.1.tcl (hunk 1 above)
Comment (mascguy) [reply _resources/port1.0/group/cmake-1.1.tcl]: > The cmake Portfile already specifies `installs_libs no` so using `depends_skip_archcheck` for cmake here should not be necessary.

Fixed, along with your other two related comments.

**Comment by pmetzger** [comment]
> This is not correct. NDEBUG is part of the C standard with a very specific meaning. DEBUG is not.

Correct. @szhorvat please submit a separate pull request.

**Comment by mascguy** [comment]
> > This is not correct. NDEBUG is part of the C standard with a very specific meaning. DEBUG is not.
> 
> Correct. @szhorvat please submit a separate pull request.

I don't think that's necessary. This PR is very debug-centric, and we could handle it here.

**Comment by mascguy** [comment]
> > This is not correct. NDEBUG is part of the C standard with a very specific meaning. DEBUG is not.
> 
> Correct. @szhorvat please submit a separate pull request.

Fixed via a subsequent commit, so we're all set. @szhorvat, any other concerns?

**Comment by szhorvat** [comment]
No, I'll leave the PR to you @mascguy unless you all have an agreement that I should submit a separate change.

**Comment by mascguy** [comment]
Oh, just noticed one other item that needs to be fixed: Flag `-DNDEBUG` needs to be added to the full suite of `configure.*flags` we handle, vis-a-vis manipulating the debug/optimization flags. Will try to fix that this evening, as it's time to return to my day job.

**Comment by RJVB** [comment]
>This is not correct. `NDEBUG` is part of the C standard with a very specific meaning. `DEBUG` is not. https://en.cppreference.com/w/c/error/assert

The fact that a token is part of the standard or not means squat for how people use it.

**Comment by RJVB** [comment]
>> The cmake Portfile already specifies `installs_libs no` so using `depends_skip_archcheck` for cmake here should not be necessary.

Does that mean `installs_libs` is now actually used for that kind of thing? I thought the token had become obsolete itself because of I forget which new addition.

**Comment by pmetzger** [comment]
> I don't think that's necessary. This PR is very debug-centric, and we could handle it here.

It hasn't been handled for over a year and frankly I think the odds of it ever converging on a completed PR are very low. I think this is a simple change and I would like it to be incorporated. @szhorvat if you're not willing to submit it, let me know, we could always handle it internally.

**Comment by mascguy** [comment]
> > I don't think that's necessary. This PR is very debug-centric, and we could handle it here.
> 
> It hasn't been handled for over a year and frankly I think the odds of the whole of this ever converging on a completed PR are very low. I think this is a simple change and I would like it to be incorporated. @szhorvat if you're not willing to submit it, let me know, we could always handle it internally.

I have this covered. So let me handle it, please.

**Comment by mascguy** [comment]
Observers: Logistics-wise, even after we have consensus, this isn't done per se: Given the criticality of this portgroup, I want to test locally with a swath of various ports, to ensure we don't have any breakage. (Both with, and without, `+debug`.)

That shouldn't take too long, but I need to carve out a few dedicated hours.

So best-case, we might - might - be able to merge by Midnight tonight. But that's only if no issues are encountered.

**Comment by mascguy** [comment]
So... I've been working with the changes from the PR for the last two hours, particularly focused on the debug support. (Specifically ensuring that pg `cmake` closely mirrors what we're currently doing, in pg `debug`.)

And then it dawned on me: Why duplicate the same logic, when we can simply include pg `debug` within pg `cmake`?

This provides the best of all worlds:
* Absolutely consistent debug variant behavior - and debug-related features - without having to copy-pasta logic between the two portgroups;
* Simplification of pg `cmake`, as everything debug-related is delegated to pg `debug`;
* Additional debug-related features, like variant `debugoptimized`. (The latter was a recent addition to pg `debug`, made three months ago.)
* Etc.

The only thing not currently supported by pg `debug`, is the ability to modify the debug flags used. (Which would be necessary to support the concept of new option `cmake.debugopts`.) But that could easily be added, if there's demand. And as it turns out, Yours Truly is the defacto maintainer of pg `debug`, as I completely overhauled it one year ago. So no worries there.

In any event, more thought and pondering is needed, over the next day or two.

And finally, several hours were spent on moving this along today, and the plan is to finalize our approach over the next few days.

Stay tuned!

p.s. In the interim, this PR should remain open: It's better to get this right from the start, rather than trying to rush it along and cause mayhem...

**Comment by mascguy** [comment]
Initial round of changes committed, re: delegating debug variant handling to pg `debug`.

Initial testing looks promising, but much more is needed. Thus, this is still very much a work-in-progress.

**Comment by mascguy** [comment]
FYI, there was one bug found during testing, related to the initial commit: The new code added to support `samu` - in proc `cmake::handle_generator` - references `prefix`. But that wasn't added to the global declarations at the top of that proc, which caused a fatal error.

I also squashed the various changes down to two commits: The original few from Rene are encompassed in the first, with my changes to utilize pg `debug` in the second.

Otherwise, things are looking good so far. Still needs more testing, but making progress...

**Comment by mascguy** [comment]
Observers: In terms of moving forward with this, we'll ultimately want to pull in others to review and comment. (Though we're nowhere near there yet, as my testing thus far has been cursory at best.)

But given how critical this portgroup is, reat assured that we'll do everything possible - as much as is practical - to ensure eagle-eyed folks have an opportunity to double-check our work. Ditto for raising potential edge cases and such, that we need to test for.

Once I've done more in-depth testing, I'll ask folks to test locally. But don't waste any time on that yet, unless you're feeling adventurous.

More to come, over the next few days...

**Comment by RJVB** [comment]
You make it sound as if you've been overhauling this file completely, was that necessary?! (and I am indeed only observing the email notifications I get...)

**Comment by mascguy** [comment]
> You make it sound as if you've been overhauling this file completely, was that necessary?! (and I am indeed only observing the email notifications I get...)

Understandable, given my comments.

But in reality, the changes are limited to two things:
* Your additions for supporting Samurai, essentially verbatim. (Plus the one-line fix for accessing `prefix` from the global scope.)
* Stripping out debug support, in favor of including pg `debug`.

Otherwise, there's no plan to touch anything else, unless deemed necessary by all of you folks.

**Comment by mascguy** [comment]
@lukaso Lukas, you'll be interested in this for GIMP, as it extends the use of pg `debug` to all ports using `cmake-1.1`. (Which covers most of our CMake-based ports, apart from a few still based on `cmake-1.0`.)

**Comment by lukaso** [comment]
> @lukaso Lukas, you'll be interested in this for GIMP, as it extends the use of pg `debug` to all ports using `cmake-1.1`. (Which covers most of our CMake-based ports, apart from a few still based on `cmake-1.0`.)

That's great!

**Comment by pmetzger** [comment]
This "draft" is now 2.5 years old. The PR queue is not intended to be either a mortuary or a retirement home. Should anyone wish to actually finish working on this, having this ticket closed will not inhibit their efforts in any way; this PR can be reopened at will or a new one can be created once someone has something that can be merged that they wish to pull.

**Comment by mascguy** [comment]
I'll reopen this at some point, once I've had a chance to thoroughly test.
